### ISSUE TITLE
Illegal Buffer when trade is processed.

### ISSUE THREAD

**Issue by markrmessmore** [Arcana/node-dota2#507]
Hi all. To be up front, I believe this is the same error as listed [here](https://github.com/Arcana/node-dota2/issues/480). 

Given I've been coding a whopping three months I can tell that ssovit and @Crazy-Duck resolved the issue however I'm having trouble following how the resolution was implemented (not only the code they used but where in the process it should go, etc...).

I see @Crazy-Duck mentioned a branch called 'items' but that seems to no longer exist. I'm using version 5.1.0 of node-dota2 so I guess my first question is...is this the same error showing up for a different reason? Has a 'full fix' not been implemented? How do I resolve my own Illegal Buffer? Why is monosyllable such a long word? That sort of stuff. I've attached the pertinent files for my code and the error.
[dotaBot.txt](https://github.com/Arcana/node-dota2/files/1065898/dotaBot.txt)
[dotaTrades.txt](https://github.com/Arcana/node-dota2/files/1065899/dotaTrades.txt)
[error.txt](https://github.com/Arcana/node-dota2/files/1065897/error.txt)

Any help/guidance would be appreciated.

**Comment by Crazy-Duck** [comment]
That error was merged into the main repository yes and should be in 5.1.0... Can you verify if you're actually on the latest version of the library?

**Comment by markrmessmore** [comment]
Can confirm: 

`
C:\Users\markr\Google Drive\Personal\WLDL\wldleague> npm list dota2
wldleague-bunnybot@0.6.0 C:\Users\markr\Google Drive\Personal\WLDL\wldleague
-- dota2@5.1.0
`

**Comment by markrmessmore** [comment]
FWIW I manually processed a trade while the bot was on, but it had no part in the transaction. Upon receiving the updated inventory snapshot I again got the same error as above.

**Comment by Crazy-Duck** [comment]
Looks to me as if there's an item in your inventory that isn't getting parsed correctly... Which is incredibly hard to debug since, well, I don't have your inventory :/ If you know a bit of node js , you could try to put a breakpoint at line 37 of handlers/cache.js and look at the contents of obj_data. Normally that should be an array of byte arrays iirc that can be decoded as CSOEconItems. If you step through it, you should be able to determine on what item it errors out

**Comment by markrmessmore** [comment]
I'll be glad to try and maneuver my way through it if you think that's what's going on. However I have another function that simply checks the status of my inventory:

    exports.getInventory = function (userid, settings, Discord, discordClient) {
      tradeManager.getInventoryContents(570, 2, true, (err, inventory) => {
        if (err) {
          console.log(err);
        } else {
          tradeManager.getOffers(1, function (err, sent, received){
            if (err) {
              console.log(err);
            }
            else {
              var pendingGifts = sent.length;
              console.log(pendingGifts);
              pushInventory.pushInventory(userid, settings, Discord, discordClient, inventory, pendingGifts);
            }
          });
        }
      });
    };

That isn't erring whatsoever. Perhaps I'm misunderstanding and you're speaking of something different. However I guess I'm uncertain why the inventory would error at one part and not another.

**Comment by Crazy-Duck** [comment]
That function uses other means to parse the inventory (at a first glance I'd say that's node-steam) so it's very much possible that that one works whereas ours errors out. I'd very much appreciate it if you could do a debug trace to determine which item bugs out the parser

**Comment by markrmessmore** [comment]
Ok so I've tried adding both:
    
    console.log(object_data);
    break;

and

    console.trace();

at line 37 and I'm coming up with the same error. On initial start up I'll have data appear on my console however when trying to process the trade the error will happen. Did I test as you were requesting or did I misunderstand?

It's also worth noting that I have both a production and test environment (i.e. two steam accounts, therefore two different steam inventories) and they're both responding the same way. Therefore I don't believe it boils down to an item erring out.

Thanks for any help you can offer.

**Comment by jimmydorry** [comment]
I'm guessing that @Crazy-Duck needs a dump of your items or for you to isolate which item causes the error. 

As @Crazy-Duck can't reproduce it, he can't debug it.

**Comment by markrmessmore** [comment]
I'm fine doing that but given my n00bishness - will that just require a console.log of the object_data?

Also - if the belief is that it is specific to an item, why would this be happening in two separate inventories? Their items are completely different.

**Comment by Crazy-Duck** [comment]
There might be more than one that crashes tbh, there's a lot of them :)
As for the debugging, replace L39 in cache.js 
```
var items = object_data.map(obj => Dota2.schema.lookupType("CSOEconItem").decode(obj));
```
by
```
var items = object_data.map(obj => {
    this.Logger.debug(obj);
    Dota2.schema.lookupType("CSOEconItem").decode(obj);
});
```
That should print out every object before attempting the decode

**Comment by markrmessmore** [comment]
Ok I replaced line 37 (I was assuming that's what you meant instead of 39 since you mentioned it earlier and it matched the code). This section of code now is as follows:

    function handleSubscribedType(obj_type, object_data, isDelete) {
        switch (obj_type) {
            // Inventory item
            case cacheTypeIDs.CSOEconItem:
                this.Logger.debug("Received inventory snapshot");
               
                var items = object_data.map(obj => {
                  this.Logger.debug(obj);
                  let parsed = Dota2.schema.lookupType("CSOEconItem").decode(obj);
                  this.Logger.debug(JSON.stringify(parsed));
                  return parsed;
                });

                var inv = this.Inventory.filter(item => items.reduce((acc, val) => acc && item.id.notEquals(val.id)), true);
                if (!isDelete) inv.concat(items);
                this.emit("inventoryUpdate", inv);
                this.Inventory = inv;
                break;

I'm getting the same result. The error is as follows:

```
Logged into Steam
3 July 20:06:30 - Launching Dota 2
Connection to MongoDB verified!
BunnyBot (Discord) reporting for duty!
3 July 20:06:31 - Sending ClientHello
3 July 20:06:31 - Received client welcome.
3 July 20:06:31 - Joining chat channel: Elihu
Dota2 ready.
3 July 20:06:31 - Cache(s) subscribed, type(s): 1,7
3 July 20:06:31 - Received inventory snapshot
3 July 20:06:31 - <Buffer 08 b2 a3 aa d0 28 10 bf ec c2 c5 01 18 83 80 80 80 0c 20 c0 5d 38 04 40 04 48 1f 78 00 80 01 e8 90 b9 da 27>
3 July 20:06:31 - {"id":"10905883058","account_id":414234175,"inventory":3221225475,"def_index":11968,"quality":4,"flags":4,"origin":31,"style":0,"original_id":"10658662504"}
3 July 20:06:31 - <Buffer 08 bc e7 b7 d0 28 10 bf ec c2 c5 01 18 83 80 80 80 0c 20 bd 5d 38 04 40 04 48 1f 62 09 08 90 03 1a 04 02 00 00 00 78 01 80 01 f9 aa bc 82 27>
3 July 20:06:31 - {"id":"10906104764","account_id":414234175,"inventory":3221225475,"def_index":11965,"quality":4,"flags":4,"origin":31,"attribute":[{"def_index":400,"value_bytes":"AgAAAA=="}],"style":1,"original_id":"10474165625"}
3 July 20:06:31 - <Buffer 08 dd f6 e3 f4 29 10 bf ec c2 c5 01 18 83 80 80 80 0c 20 b0 5b 38 04 40 04 78 00 80 01 fb 8b ed f9 1b>
3 July 20:06:31 - {"id":"11250760541","account_id":414234175,"inventory":3221225475,"def_index":11696,"quality":4,"flags":4,"style":0,"original_id":"7503300091"}
3 July 20:06:31 - <Buffer 08 f3 be c2 f6 29 10 bf ec c2 c5 01 18 83 80 80 80 0c 20 cc 5d 38 04 40 04 48 22 78 00 80 01 b0 ec fa db 26>
3 July 20:06:31 - {"id":"11254407027","account_id":414234175,"inventory":3221225475,"def_index":11980,"quality":4,"flags":4,"origin":34,"style":0,"original_id":"10393400880"}
3 July 20:06:31 - <Buffer 08 92 9d 8e 8e 2a 10 bf ec c2 c5 01 18 83 80 80 80 0c 20 ba 59 38 04 40 04 78 00 80 01 ab ee d5 b8 19>
3 July 20:06:31 - {"id":"11303882386","account_id":414234175,"inventory":3221225475,"def_index":11450,"quality":4,"flags":4,"style":0,"original_id":"6829733675"}
3 July 20:06:31 - <Buffer 08 e3 9f 98 8e 2a 10 bf ec c2 c5 01 18 83 80 80 80 0c 20 c8 53 38 04 40 04 48 22 78 00 80 01 c2 90 cd f0 27>
3 July 20:06:31 - {"id":"11304046563","account_id":414234175,"inventory":3221225475,"def_index":10696,"quality":4,"flags":4,"origin":34,"style":0,"original_id":"10705127490"}
3 July 20:06:31 - <Buffer 08 f7 90 9e 8e 2a 10 bf ec c2 c5 01 18 83 80 80 80 0c 20 d7 58 38 04 40 04 78 00 80 01 db 8d c5 a6 1d>
3 July 20:06:31 - {"id":"11304142967","account_id":414234175,"inventory":3221225475,"def_index":11351,"quality":4,"flags":4,"style":0,"original_id":"7865452251"}
3 July 20:06:31 - <Buffer 08 b1 82 d3 95 2a 10 bf ec c2 c5 01 18 83 80 80 80 0c 20 a4 59 38 04 40 04 78 00 80 01 93 9f a2 81 1e>
3 July 20:06:31 - {"id":"11319689521","account_id":414234175,"inventory":3221225475,"def_index":11428,"quality":4,"flags":4,"style":0,"original_id":"8055721875"}
3 July 20:06:31 - <Buffer 08 df 80 d4 95 2a 10 bf ec c2 c5 01 18 83 80 80 80 0c 20 c7 53 38 04 40 04 48 22 62 09 08 b9 01 1a 04 18 67 c7 57 62 09 08 ba 01 1a 04 fb 38 00 06 78 ... >
3 July 20:06:31 - {"id":"11319705695","account_id":414234175,"inventory":3221225475,"def_index":10695,"quality":4,"flags":4,"origin":34,"attribute":[{"def_index":185,"value_bytes":"GGfHVw=="},{"def_index":186,"value_bytes":"+zgABg=="}],"style":0,"original_id":"9725132571"}
3 July 20:06:31 - <Buffer 08 ad 99 d4 95 2a 10 bf ec c2 c5 01 18 83 80 80 80 0c 20 dd 5b 38 04 40 04 78 00 80 01 cd 96 c5 a6 1d>
3 July 20:06:31 - {"id":"11319708845","account_id":414234175,"inventory":3221225475,"def_index":11741,"quality":4,"flags":4,"style":0,"original_id":"7865453389"}
3 July 20:06:31 - <Buffer 08 bc b2 d4 95 2a 10 bf ec c2 c5 01 18 83 80 80 80 0c 20 cc 55 38 04 40 04 48 22 78 00 80 01 b8 90 b8 e9 24>
3 July 20:06:31 - {"id":"11319712060","account_id":414234175,"inventory":3221225475,"def_index":10956,"quality":4,"flags":4,"origin":34,"style":0,"original_id":"9884796984"}
3 July 20:06:31 - Unhandled cache ID: 7
3 July 20:06:32 - Chat channel Elihu has 0 person(s) online
Ready to Trade!
Trade request received
3 July 20:07:41 - Create, type 1
3 July 20:07:41 - Received inventory snapshot
3 July 20:07:41 - 8
C:\Users\Messmore\Google Drive\Personal\WLDL\wldleague\node_modules\protobufjs\src\reader.js:47
        throw Error("illegal buffer");
        ^

Error: illegal buffer
    at create_typed_array (C:\Users\Messmore\Google Drive\Personal\WLDL\wldleague\node_modules\protobufjs\src\reader.js:47:15)
    at Function.create_buffer [as create] (C:\Users\Messmore\Google Drive\Personal\WLDL\wldleague\node_modules\protobufjs\src\reader.js:69:19)
    at Type.CSOEconItem$decode [as decode] (eval at Codegen (C:\Users\Messmore\Google Drive\Personal\WLDL\wldleague\node_modules\@protobufjs\codegen\index.js:50:33), <anonymous>:5:12)
    at object_data.map.obj (C:\Users\Messmore\Google Drive\Personal\WLDL\wldleague\node_modules\dota2\handlers\cache.js:40:67)
    at Buffer.map (native)
    at Dota2Client.handleSubscribedType (C:\Users\Messmore\Google Drive\Personal\WLDL\wldleague\node_modules\dota2\handlers\cache.js:38:37)
    at Dota2Client.onCreate (C:\Users\Messmore\Google Drive\Personal\WLDL\wldleague\node_modules\dota2\handlers\cache.js:205:26)
    at SteamGameCoordinator.fromGC (C:\Users\Messmore\Google Drive\Personal\WLDL\wldleague\node_modules\dota2\index.js:167:38)
    at emitThree (events.js:116:13)
    at SteamGameCoordinator.emit (events.js:197:7)
    at SteamGameCoordinator.handlers.(anonymous function) (C:\Users\Messmore\Google Drive\Personal\WLDL\wldleague\node_modules\steam\lib\handlers\game_coordinator\index.js:107:10)
    at SteamGameCoordinator.<anonymous> (C:\Users\Messmore\Google Drive\Personal\WLDL\wldleague\node_modules\steam\lib\handlers\game_coordinator\index.js:19:28)
    at emitThree (events.js:121:20)
    at CMClient.emit (events.js:197:7)
    at CMClient._netMsgReceived (C:\Users\Messmore\Google Drive\Personal\WLDL\wldleague\node_modules\steam-client\lib\cm_client.js:278:8)
    at CMClient.handlers.(anonymous function) (C:\Users\Messmore\Google Drive\Personal\WLDL\wldleague\node_modules\steam-client\lib\cm_client.js:386:8)
[nodemon] app crashed - waiting for file changes before starting...
```

The initial startup displays the inventory. Once ```Ready to Trade``` has displayed...startup is done and it is ready to trade. You can then see what will show up the moment I open my mobile authenticator and confirm the trade when it displays: ```Trade Request Received```. Then it receives the inventory snapshot and doesn't print out the inventory again.

**Comment by Crazy-Duck** [comment]
How many items did you trade here? one or multiple? I'm gonna hazard a guess and say 1, since it looks like it's treating a single item buffer as an array of items (the 8 print seems like the first 0x08 byte all items have in common)... It looks like i'm gonna have to treat the create cache event differently than the subscribe events.. Could you try and trade 2 items and tell me what happens then?

**Comment by markrmessmore** [comment]
One indeed. I tried to trade six items with the same result. I've attached a text file with the output in case you want to look for any differences.
[error.txt](https://github.com/Arcana/node-dota2/files/1136863/error.txt)

**Comment by Crazy-Duck** [comment]
I've created a new inventory branch; it should print out the trade request. Could you try it out and tell me what it spits out?

**Comment by markrmessmore** [comment]
I *think* I did this properly. I downloaded the zip, renamed my dota2 file in my node-modules to dota2.sav and pasted this folder in as dota2. If those are not the proper steps (remember I'm a n00b) please let me know. 

I initially got a 'this is not a function' so I'm *guessing* you wanted lines 37 and 39 to read ```this.Logger.debug``` where the ```.debug``` was missing. If in fact that is accurate, the trade did process without error. Here is the output:

```
Connection to MongoDB verified!
Logged into Steam
2 July 09:31:24 - Launching Dota 2
BunnyBot (Discord) reporting for duty!
2 July 09:31:25 - Sending ClientHello
2 July 09:31:26 - Received client welcome.
Dota2 ready.
2 July 09:31:26 - Cache(s) subscribed, type(s): 1,7
2 July 09:31:26 - Received inventory snapshot
2 July 09:31:27 - Unhandled cache ID: 7
Ready to Trade!
Trade request received
2 July 09:32:50 - Create, type 1
2 July 09:32:50 - Received trade request
2 July 09:32:50 - {"id":"11891042381","account_id":414234175,"inventory":3221225475,"def_index":11513,"quality":4,"flags":4,"origin":8,"style":0,"original_id":"8835450347"}
2 July 09:32:50 - Create, type 1
2 July 09:32:50 - Received trade request
2 July 09:32:50 - {"id":"11891042441","account_id":414234175,"inventory":3221225475,"def_index":11093,"quality":4,"flags":4,"origin":34,"style":0,"original_id":"10512540967"}
```

**Comment by Crazy-Duck** [comment]
You can also update your dependency in your package.json to read `dota2: Arcana/node-dota2#inventory` and do an npm update after iirc. That way you don't need to mess around with downloading zips and all :) About the missing debug, yeah, I messed that up, good catch and fix ;) I've fixed it on the branch as well in the mean time.
Now, for your output, is the first trade request for 1 item and the second one for 2 items?

**Comment by Crazy-Duck** [comment]
Oh and what happens when you accept the trade would be useful too

**Comment by markrmessmore** [comment]
The output above is a single trade request with two items to ensure multiple things didn't throw a wrench into the result. After the trade is accepted no additional pieces show up except for ```Trade accepted``` which I've told it to do.

**Comment by Crazy-Duck** [comment]
Oh ok, so you receive a cache creation event for each item in the trade request, that makes sense. Is there any additional output when you accept/refuse the trade request? Like are there cache destroy events arriving?

**Comment by markrmessmore** [comment]
Because this is really a donation function it auto-accepts and only denies if something is asked for in return. In either case there is no additional output.

**Comment by Crazy-Duck** [comment]
Wait, that actually sucks.. So there is no way to determine the trade went through?? Could you do the test with debugmore activated to see if there's really nothing at all?

**Comment by markrmessmore** [comment]
To this point I've been seeing if the items ended up in their intended locations to see if the trade processed. And thus far they have been (even when the illegal buffer error started). To simplify things I've added a couple ```console.log``` responses to clarify. Here are two transactions the first accepted, the second declined:

```
Trade request received
6 July 09:01:35 - Create, type 1
6 July 09:01:35 - Received trade request
6 July 09:01:35 - {"id":"11948521246","account_id":414234175,"inventory":3221225475,"def_index":10023,"quality":4,"flags":4,"origin":34,"attribute":[{"def_index":185,"value_bytes":"jxoEWA=="},{"def_index":186,"value_bytes":"e3WeCg=="}],"style":0,"original_id":"9639100834"}
6 July 09:01:35 - Create, type 1
6 July 09:01:35 - Received trade request
6 July 09:01:35 - {"id":"11948521337","account_id":414234175,"inventory":3221225475,"def_index":11099,"quality":4,"flags":4,"origin":31,"style":0,"original_id":"8114337104"}
Donation accepted.
```

```
Trade request received
Trade declined: user was asking for items in return.
```
I tried replacing ```debug``` with ```debugmore``` on lines **37** & **39** however I received an error saying that it wasn't a function. Not sure if I did that correctly or not.

**Comment by Crazy-Duck** [comment]
The debugmore is when you create the dota2 object. The constructor takes 3 arguments:
```
Dota2.Dota2Client = function Dota2Client(steamClient, debug, debugMore) {
    ...
}
```
Right now I assume you initialize by only passing a `steamClient` object and `true`, try passing a third argument `true` additionally as well. It will log additional stuff then.

So the cache create events show up when you've accepted the trade? And when you decline the trade, you get no cache messages? Alright, that makes sense then.

**Comment by markrmessmore** [comment]
I follow you now. Thanks for the explanation.

Not a ton more information passed with debugMore. Here's the output on a successful trade:
```
0 July 12:12:23 - Dota2 fromGC: k_ESOMsg_Create
0 July 12:12:23 - Create, type 1
0 July 12:12:23 - Received trade request
0 July 12:12:23 - {"id":"11968472008","account_id":414234175,"inventory":3221225475,"def_index":11507,"quality":4,"flags":4,"origin":8,"style":0,"original_id":"8568430760"}
0 July 12:12:24 - Dota2 fromGC: k_ESOMsg_Create
0 July 12:12:24 - Create, type 1
0 July 12:12:24 - Received trade request
0 July 12:12:24 - {"id":"11968472067","account_id":414234175,"inventory":3221225475,"def_index":11505,"quality":4,"flags":4,"origin":8,"style":0,"original_id":"8835450346"}
Donation accepted.
```

**Comment by Crazy-Duck** [comment]
Alright, and as for my last question? Can you confirm that the create lines only show up after you've accepted the trade?

**Comment by markrmessmore** [comment]
Yes. The create lines only show up after the trade has been accepted. Otherwise it sits at 'trade request received.'

**Comment by Crazy-Duck** [comment]
Alright, sorry for taking so long, but I've finally gotten around to updating the inventory branch. Your inventory should now be properly updated after processing a trade. Could you test and let me know if everything works as it should?

**Comment by markrmessmore** [comment]
Just so I'm clear before I do this - you're asking me to download the updates specifically from the ***Inventory*** branch (not the standard). Update my bot with said data. Test and report back.

**Comment by Crazy-Duck** [comment]
Yup, that's it :)

**Comment by markrmessmore** [comment]
Everything went through great. Two trades are reflected below. The first was just one item. The second was three items:

```
Ready to Trade!
Trade request received
4 August 15:28:52 - Create, type 1
4 August 15:28:52 - Got an item traded
Donation accepted.
Trade request received
4 August 15:30:09 - Create, type 1
4 August 15:30:09 - Got an item traded
4 August 15:30:09 - Create, type 1
4 August 15:30:09 - Got an item traded
4 August 15:30:10 - Create, type 1
4 August 15:30:10 - Got an item traded
Donation accepted.
```

**Comment by Crazy-Duck** [comment]
Alright, thanks a lot for the help!

**Comment by markrmessmore** [comment]
No...thank you! :-)
